### ISSUE TITLE
Discussion on how to add sinks to `adapt()`

### ISSUE THREAD

**Issue by Evertt** [state-adapt/state-adapt#83]
So you mentioned you were planning to add sinks to `adapt()` and I was just thinking about what that could / should look like. And I was thinking that of course, ideally, you would want to be able to only run sinks / side-effects when precisely the state-properties they depend on have changed.

Like this for example:

```ts
const user = adapt(
    {
        name: "Bob",
        age: 20,
    },
    {
        sinks: [
            // This side-effect should only run when age changes, not when name changes.
            // Otherwise the user would receive a happy birthday message way too often.
            ({ age }) => age >= 21 && alert("Happy birthday! You can drink now!"),
        ],
    },
)
```

Which makes me think that the `state` object that's given to sinks should be a relatively simple `Proxy` object that keeps track of which properties on it are read to determine the dependencies of the sink. That should not be too hard, but that would only work for top-level properties.

~~If you'd want to track deeper nested properties then I imagine you'd need to offer a fully signal-based deeply reactive proxy object that turns every property at every depth level into a (derived) signal. But then those signals will need to be used with syntax like `user.address.street.value` or `user.address.street()`, neither of which is ideal.~~

To track deeper nested properties I would just recommend that selectors are used for that. Like this:

```ts
const user = adapt(
    {
        name: "Bob",
        age: 20,
        address: {
            street: "Boulevard Avenue",
        },
    },
    {
        selectors: {
            street: user => user.address.street,
        },
        sinks: [
            // Now `street` should be available at the top-level of the `state` object,
            // and therefor trackable, because it was declared as a selector.
            state => console.log(state.street)
        ],
    },
)
```

As you can see for now I'm imagining sinks to be a simple array of functions, because I can't imagine what else they could be.

What do you think? Any feedback on this?

**Comment by mfp22** [comment]
Those were mostly my thoughts exactly, except for the proxy thing. You can just give it a selector for all cases. 

What I'm wondering is if sinks should subscribe independently. Or if they should follow the lifecycle of the state as it's consumed in the UI. I can think of situations for both. So maybe both should be supported somehow.

I thought about naming them effects because that's what they would normally be named, but I'm afraid it would encourage people to use them like they use effects in other libraries. They should be thought of as places where data ends up. Not where you trigger state changes in other stores or anything like that.

**Comment by Evertt** [comment]
> You can just give it a selector for all cases.

I don't really understand what you mean with this. Do you mean that `age` would also only be available to the sink if it were declared as a selector? I mean, I'd call that less than ideal, but of course that's an option. But then still you'd want to know *which* selectors were used by a particular sink in order to determine when the sink should be run next time, no?

> What I'm wondering is if sinks should subscribe independently. Or if they should follow the lifecycle of the state as it's consumed in the UI.

I don't understand how sinks could have an independent lifecycle? Could you explain? I mean, sinks only get run when state changes, right? And state can't change if nothing is subscribed to it. I wasn't actually 100% sure about that, but I just tried it and I indeed got an error saying `"Cannot apply update before store "0" is initialized"`.

> They should be thought of as places where data ends up.

I agree. That's why I vote for the term `sinks`.

> Not where you trigger state changes in other stores or anything like that.

But I'm assuming using a sink to `next()` into another `Source` is a valid use-case, right?

**Comment by mfp22** [comment]
We only need selectors. If it's inconvenient for sinks, it's probably inconvenient for the UI. For objects I usually create the adapter using `joinAdapters` and that creates selectors for each property automatically. This is why I didn't already implement automatically creating selectors for object states. And that's something you'd want at the `createAdapter` level too, but you can't do it without an actual state object to iterate over, and adapters don't have that. Maybe they could take in a `defaultState` parameter in the first function call, but that shouldn't be required, and TypeScript wouldn't know how `State` is being inferred so it couldn't detect whether selectors are created automatically or not. So I just use `joinAdapters`. It usually ends up preventing the `....` boilerplate anyway. 

I was wondering if some sinks should subscribe and cause the state to exist. If I made a `localStorage` sink and people pass it in, they might expect that alone to cause the store to always be there. The DOM is a sink technically, and when it subscribes it causes the state to exist, so why wouldn't other sinks do the same?

Actually `subscribe` itself can be used to define these kinds of sinks. For example, let's say a route of the application needs some sound to be played. I know there's an `audio` element, but let's say it's not enough or there is some output that has to be done in JavaScript. So there is no DOM representation of the side-effect, but it should be thought of as an important output of state—a reason for the state to exist. In this situation, you can just `audioStore.state$.subscribe(state => { doTheAudioStuff })` and have it clean up when the route is destroyed too. 

But one thing that there isn't today is a way to define side-effects that tag along with the store's state as other things subscribe to it. `localStorage` is one of these. It's like a passive observer while the important application stuff manages the actual state. And this would be not too hard to implement. 

> But I'm assuming using a sink to next() into another Source is a valid use-case, right?

I can't think of any situation where this would be a good idea.

**Comment by Evertt** [comment]
Thanks for explaining, all of that clears a lot up for me.

> I can't think of any situation where this would be a good idea.

Yeah this is showing that I actually still have very little experience using StateAdapt in the wild. The main reason I thought that wiring a sink into another `Source` was because I was thinking of the diagrams that you sometimes draw when you show how data flows from sources to sinks. For example here:

<img width="164" alt="image" src="https://github.com/state-adapt/state-adapt/assets/1267282/a3ade3d8-e876-4374-8b46-32d53c882d1b">

The `paused` circle in the diagram looked to me link it was both a sink for everything above it and a source for everything underneath it. That's why I thought maybe a piece of state created with StateAdapt could also be in the middle and therefor connect to other `Source`s.

But I guess I'll only understand how wrong that is once I start using StateAdapt more in the wild.

**Comment by mfp22** [comment]
Nah, that's just the data flow. Sources are top-level, like events. Here's the code that goes with that diagram https://github.com/mfp22/ngrx-cs-stateadapt-sylomovies/blob/master/src/app/components/carousel/carousel.component.ts

**Comment by Evertt** [comment]
> Actually `subscribe` itself can be used to define [active] sinks.

Yeah, thinking about it in terms of "passive" and "active" sinks, I'm thinking:

When you want to create an active sink, you'll have to manage the subscription yourself anyway. As in, you'll need to configure yourself when that subscriber should unsubscribe. Cause even though it might often be tied to a component lifecycle, it doesn't have to be.

So considering that, I think it makes most sense to implement the sinks in `adapt({}, { sinks: [] })` as purely passive sinks. Which would have their subscriptions automatically managed by `adapt()` based on whenever `store.state$` or any other relevant selector is being subscribed to by some other sink (e.g. the DOM / template).

And whenever you'd want to use an active sink, you should just use a manual `subscribe` for that and manage that subscription yourself. Which in many cases might still be as easy as using a `takeUntilDestroyed()` operator built for your specific framework.

**Comment by mfp22** [comment]
Exactly

**Comment by mfp22** [comment]
I expect the most common use case for sinks is where a dev will import a sink someone else defined. StateAdapt can provide a `localStorageSink` as it will be extremely common. So with that in mind, I'd rather make the central integration extremely simple and put more logic/syntax on the burden of the sink plugins. I actually think just using observables might be the best idea. So they're just like `Source`s, but without using the return value of the observables. Then they can even define their own initialization logic with `defer` and teardown logic with `finalize`. 

```typescript
const localStorageSink = store => store.state$.pipe(
  tap(state => localStorage.setItem(store.__.path, JSON.stringify(state)),
);

const initialState = {
    name: "Bob",
    age: 20,
    address: {
        street: "Boulevard Avenue",
    },
};
const user = adapt(initialState, {
    sinks: [localStorageSink, consoleSink],
});
```

What do you think?

Alternatively, 

```typescript
const localStorageSink = getLocalStorageSink(adapter, 'someSelector');
```

This is just an idea. Not sure how to implement it since `store` needs to be passed into the sinks and hasn't totally been defined yet. I guess it will have been when the `defer` is running. 

And the React integration... Will devs expect the sink to be subscribed to immediately?

**Comment by mfp22** [comment]
Now I'm wondering if this could be done with the existing API.

```typescript
const user = adapt(initialState, {
    sources: store => ({
        noop: [getLocalStorageSink(store), getConsoleSink(store)],
    },
});
// vs 
const user = adapt(initialState, {
    sinks: [localStorageSink, consoleSink],
});
```

It's not where you'd expect to see side-effects... And in Redux Devtools I think this would log an action that doesn't change any states. But if `getLocalStorageSink(store)` returns an action of type `localStorage sink` or something, maybe that would actually be useful. 

The reason I added `noop` in the first place was 1. For side-effects, now that I think of it 2. To get actions to show up in Redux Devtools even if I don't have them affect state at all. I've used it to log an error from HTTP I think. 

If we just use the existing API, I wouldn't change docs, but I need to create a tutorial and I would definitely mention it there. 

I really love reusing concepts. I hate redundant APIs. I also hate the name `noop`. Maybe it should be renamed to `sink` or similar and we keep the existing functionality.

**Comment by mfp22** [comment]
![image](https://github.com/state-adapt/state-adapt/assets/13700926/d1acc288-4f2d-4cce-8a9c-e32ad9a711b8)

![image](https://github.com/state-adapt/state-adapt/assets/13700926/81afd750-39e8-4603-bf49-6dbb54294a4c)

I actually really like the idea of having the sink show up in DevTools.

**Comment by kingjordan** [comment]
If i wanted to log the navigation events in angular to the redux dev tools, would a sink be the pace to do it?

**Comment by mfp22** [comment]
For now, yes. Eventually I want to create an official router sink + adapter
